consensus/scheme/rolldpos/rolldposctx.go

```diff
@@ -0,0 +1,703 @@
+	if !ok {
+		return false
+	}
+	validNum := set.NumOfValidEndorsements(
+		expectedTopics,
+		ctx.epoch.delegates,
+	)
+	numDelegates := len(ctx.epoch.delegates)
+	return numDelegates >= 4 && validNum > numDelegates*2/3 ||
+		numDelegates < 4 && validNum >= numDelegates
+}
+
+func (ctx *rollDPoSCtx) newConsensusEvent(
+	eventType fsm.EventType,
+	data interface{},
+) *consensusfsm.ConsensusEvent {
+	return consensusfsm.NewConsensusEvent(
+		eventType,
+		data,
+		ctx.round.height,
+		ctx.round.number,
+		ctx.clock.Now(),
+	)
+}
+
+func (ctx *rollDPoSCtx) calcWaitDuration() (time.Duration, error) {
+	// TODO: Update wait duration calculation algorithm
+	// If the proposal interval is not set (not zero), the next round will only be started after the configured duration
+	// after last block's creation time, so that we could keep the constant
+	waitDuration := time.Duration(0)
+	// If we have the cached last block, we get the timestamp from it
+	duration, err := ctx.calcDurationSinceLastBlock()
+	if err != nil {
+		return waitDuration, err
+	}
+	interval := ctx.cfg.FSM.ProposerInterval
+	if interval > 0 {
+		waitDuration = (interval - (duration % interval)) % interval
+	}
+
+	return waitDuration, nil
+}
+
+func (ctx *rollDPoSCtx) loggerWithStats() *zap.Logger {
+	numProposals := 0
+	numLocks := 0
+	numCommits := 0
+	if ctx.round.block != nil {
+		blkHashHex := hex.EncodeToString(ctx.round.block.Hash())
+		endorsementSet := ctx.round.endorsementSets[blkHashHex]
+		if endorsementSet != nil {
+			numProposals = endorsementSet.NumOfValidEndorsements(
+				map[endorsement.ConsensusVoteTopic]bool{
+					endorsement.PROPOSAL: true,
+					endorsement.COMMIT:   true,
+				},
+				ctx.epoch.delegates,
+			)
+			numLocks = endorsementSet.NumOfValidEndorsements(
+				map[endorsement.ConsensusVoteTopic]bool{
+					endorsement.LOCK:   true,
+					endorsement.COMMIT: true,
+				},
+				ctx.epoch.delegates,
+			)
+			numCommits = endorsementSet.NumOfValidEndorsements(
+				map[endorsement.ConsensusVoteTopic]bool{
+					endorsement.COMMIT: true,
+				},
+				ctx.epoch.delegates,
+			)
+		}
+	}
+	return ctx.Logger().With(
+		zap.Int("numProposals", numProposals),
+		zap.Int("numLocks", numLocks),
+		zap.Int("numCommits", numCommits),
+	)
+}
+
+func (ctx *rollDPoSCtx) newEndorsement(topic endorsement.ConsensusVoteTopic) (consensusfsm.Endorsement, error) {
+	endorsement := endorsement.NewEndorsement(
+		endorsement.NewConsensusVote(
+			ctx.round.block.Hash(),
+			ctx.round.height,
+			ctx.round.number,
+			topic,
+		),
+		ctx.pubKey,
+		ctx.priKey,
+		ctx.encodedAddr,
+	)
+
+	return &endorsementWrapper{endorsement}, nil
+}
+
+func (ctx *rollDPoSCtx) isProposedBlock(hash []byte) bool {
+	if ctx.round.block == nil {
+		ctx.Logger().Error("block is nil")
+		return false
+	}
+	blkHash := ctx.round.block.Hash()
+
+	return bytes.Equal(hash, blkHash)
+}
+
+func (ctx *rollDPoSCtx) isDelegateEndorsement(endorser string) bool {
```

Comment:
This is not enough. Some one could just impersonate a delegate. Probably we should just get rid of endorser address, and stick to endorser pubkey

cc @lizhefeng